### ISSUE TITLE
Memory error when a model training  run is restarted

### ISSUE THREAD

**Issue by shigetajima** [mravanelli/pytorch-kaldi#84]
Hi,
I am having a memory issue. it happens when a model training run is "restarted".
A strange thing is that it ran successfully up to 24 epochs continuously without memory issue. But when a run was restarted after, say, several epochs, out of memory happens. Below, I restarted a run which was interrupted after 6 epochs, and OOO happens at 7th epoch. 
Does anyone know how to fix this? Thanks.
 

------------------------------ Epoch 007 / 011 ------------------------------
Training train_clean_100 chunk = 1 / 10
Traceback (most recent call last):
  File "run_exp.py", line 191, in <module>
    [data_name,data_set,data_end_index,fea_dict,lab_dict,arch_dict]=run_nn(data_name,data_set,data_end_index,fea_dict,lab_dict,arch_dict,config_chunk_file,processed_first,next_config_file)
  File "/scratch1/staji/images/pytorch_kaldi_2019-02-13/pytorch-kaldi/core.py", line 211, in run_nn
    outs_dict['loss_final'].backward()
  File "/usr/local/lib/python3.6/dist-packages/torch/tensor.py", line 102, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "/usr/local/lib/python3.6/dist-packages/torch/autograd/__init__.py", line 90, in backward
    allow_unreachable=True)  # allow_unreachable flag
RuntimeError: CUDA out of memory. Tried to allocate 92.62 MiB (GPU 0; 15.78 GiB total capacity; 2.41 GiB already allocated; 2.94 MiB free; 474.86 MiB cached)

**Comment by TParcollet** [comment]
Hum, this is quite weird. If you run the same model from scratch it runs for more than 7 epochs ?

**Comment by mravanelli** [comment]
What I noticed is that when you are close to using the full memory of the
GPU, the script can randomly stop at any iteration. However, restarting the
model should employ exactly the same memory of the first run.

Mirco

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by shigetajima** [comment]
Hi, I agree that the model should use the same memory from the previous run when a job is restarted. 

What I noticed is that if a job gets killed/stopped shortly after the run started (say, after finishing a few chunks during the first epoch), restarting a run does not cause memory error, and the run continues until the end.

But after a job runs several epochs, restarting it causes out of memory error... I am rerunning jobs to see when memory error begins to happen. Will let you know next week. Thanks.

**Comment by mravanelli** [comment]
Thank you for your feedback. I will investigate more in detail on it as
soon as I can!

**Comment by shigetajima** [comment]
Hi, I did some tests by running pytorch-kaldi code. I trained liGRU model on Librispeech corpus, and my config file is basically the same as libri_liGRU_fmllr.cfg except parameters and paths.

The code can run for many epochs if there is no interruptions (I tested up to 021 epochs, and got no errors) but out of memory error did happen if a run was "restarted" after 1 or 2 epochs.

Does anyone experience similar things?

**Comment by TParcollet** [comment]
Could you please reduce your model, run for 3 epochs, note the maximal memory consumption (you must remove increasing sequence during training, so it starts at the maximal consumption). And then reload the model at epoch 3 and see if there is a difference in memory consumption, if yes, how much ?

**Comment by shigetajima** [comment]
Hi, Thank you for your suggestions.

I turned OFF increasing the seq length parameter (increase_seq_length_train=False) and reduced the liGRU model (from 5 to 2 layers), and then did a model training from scratch. 
But I got memory error at the very beginning and no training started, but does that make sense? 
When I used the option: increase_seq_length_train=True, it would reach the MAX seq length at the 4th epoch anyway, but I've ran a training for > 10 epochs successfully before. So I don't know why memory error still happens when I turned off increasing the seq length parameter...

Anyway, I tried reducing the model size significantly and finally the training started. 

[architecture1]
...
...
ligru_lay = 50
ligru_drop = 0.2
ligru_use_laynorm_inp = False
ligru_use_batchnorm_inp = False
ligru_use_laynorm = False
ligru_use_batchnorm = True
ligru_bidir = True
ligru_act = relu
ligru_orthinit=True


And I did model training and at the END of run after 3 epochs (=epoch 002), here is the memory info:

Memory info after the run
    AveCPU   AvePages     AveRSS     MaxRSS  AveVMSize  MaxVMSize        JobID 
---------- ---------- ---------- ---------- ---------- ---------- ------------ 
 00:00.000          0      3916K   9462736K    215204K  31122472K 31421.batch  


Resumed model training. During the next (4th) epoch 003 and chunk = 1/10, here is the memory info:

------------------------------Epoch 003 / 003 ------------------------------
Training train_clean_100 chunk = 1 / 10

    AveCPU   AvePages     AveRSS     MaxRSS  AveVMSize  MaxVMSize        JobID 
---------- ---------- ---------- ---------- ---------- ---------- ------------ 
 03:08.000       234K      5022M      5022M  22741144K  26242512K 31446.batch  


At the end of run after Epoch 003:
    AveCPU   AvePages     AveRSS     MaxRSS  AveVMSize  MaxVMSize        JobID 
---------- ---------- ---------- ---------- ---------- ---------- ------------ 
 00:00.000          0      3916K   9163148K    215204K  29330252K 31446.batch 

Do you see any problems with memory consumption? Thanks so much!

**Comment by mravanelli** [comment]
Hi,
can I ask you how much memory your GPU has? Also, it is a dedicated GPU or
the same GPU that you use for graphic visualization?
In the latter case, the randomness could be due to that (this happens for
instance when I run some toy experiments on my laptop).

Mirco

PS:
If you set increasing the seq length parameter to false, you don't split
the sentences and you process all the time steps of the speech sequence
since the beginning. It i normal that in this case the algorithm needs more
memory since the beginning.

**Comment by shigetajima** [comment]
Thanks for your time to look into this.

GPU has 16GB memory, and it is a dedicated one.

**Comment by mravanelli** [comment]
which is the largest model that you can train without memory issues?  With
16 GB you should be able to learn a big model without memory issues..

**Comment by shigetajima** [comment]
Hi, I tested a small liGRU model with 1 layer and ligru_lay=50. I ran the code and train for a few epochs, and restarted with  no errors. and I continued doing train/restart until 30 epochs with no errors.
But OOM happened when I trained a bigger liGRU model (1 layer and ligru_lay=550).

By the way I am talking to a local expert since there might be an issue about memory allocation on the GPU system I am using. I will keep you updated.
